### PR TITLE
Refactor startup scripts, envvars for all omero parameter

### PR DESCRIPTION
**Pull request by manics**
Refactor the startup-scripts for OMERO.server:
- All omero config properties can be set with environment variables
- Database setup is separate from starting OMERO.server
- Removed OMERO.grid support- this will be added back in a separate dependent repo
- Use [dumb-init](https://github.com/Yelp/dumb-init)
  - @dpwrussell In theory this means you could start a backgrounded script which polls localhost to check whether the server is ready before creating a public user, before the final `99-run.sh` script is run, avoiding the requirement to restart the server.

### CODE REVIEW

File: 50-config.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# Variable names should replace "." with "_" and "_" with "__"
+# E.g. CONFIG_omero_web_public_enabled=false
+
+import os
+from subprocess import call
+from re import sub
+
+
+for (k, v) in os.environ.iteritems():
+    omero = '/opt/omero/server/OMERO.server/bin/omero'
```
Comment (dpwrussell) [inline 50-config.py]: Need not be in loop

Comment (dpwrussell) [inline 50-config.py]: and probably should be caps

Comment (manics) [inline 50-config.py]: Fixed

**Comment by manics** [comment]
I've renamed run-exec.sh to entrypoint.sh

**Comment by dpwrussell** [comment]
Works as expected.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+Then run OMERO.server passing the the database configuration parameters if they differ from the defaults:
+
+    docker run -d --name omero-server --link postgres:db
+        -e CONFIG_omero_db_user=postgres \
+        -e CONFIG_omero_db_pass=postgres \
+        -e CONFIG_omero_db_name=postgres \
+        -e ROOTPASS=omero-root-password \
+        -p 4063:4063 -p 4064:4064 \
+        -e ROOTPASS=omero openmicroscopy/omero-server
```
Comment (joshmoore) [inline README.md]: ROOTPASS twice?

File: Dockerfile [hunk 3]
```diff
@@ -12,13 +12,15 @@ RUN yum -y install epel-release \
 ARG OMERO_VERSION=latest
```
Comment (joshmoore) [inline Dockerfile]: If we're using the main release version (i.e. `5.3.2`) for tags of this repo, then I'm assuming:

 * `latest` images are considered worst-practice
 * on each OMERO release, we'll branch from `master` and bump this string to include the version (`5.3.3`)
 * for bug fixes before an OMERO release, we'll suffix `5.3.3-1`

Note: this doesn't yet give us a strategy for breaking changes in the image _itself_

File: 60-database.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+omero=/opt/omero/server/OMERO.server/bin/omero
+omego=/opt/omero/omego/bin/omego
+cd /opt/omero/server
+
+CONFIG_omero_db_host=${CONFIG_omero_db_host:-}
+if [ -n "$CONFIG_omero_db_host" ]; then
+    DBHOST="$CONFIG_omero_db_host"
+else
+    DBHOST=db
+    $omero config set omero.db.host "$DBHOST"
```
Comment (joshmoore) [inline 60-database.sh]: Does this need to be outside the if block?

File: 50-config.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python
+# Set omero config properties from CONFIG_ envvars
+# Variable names should replace "." with "_" and "_" with "__"
```
Comment (joshmoore) [inline 50-config.py]: Not particularly worried but there's obviously some danger of a `__` conflict here.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+Example with named volumes
+--------------------------
+
+    docker volume create --name omero-db
+    docker volume create --name omero-data
+
+    docker run -d --name postgres -e POSTGRES_PASSWORD=postgres \
+        -v omero-db:/var/lib/postgresql/data postgres
+    docker run -d --name omero-server --link postgres:db
+        -e CONFIG_omero_db_pass=dbpassword  -v omero-data:/OMERO \
```
Comment (joshmoore) [inline README.md]: This example isn't working for me:

```
FATAL:  password authentication failed for user "omero"
DETAIL:  Connection matched pg_hba.conf line 95: "host all all 0.0.0.0/0 md5"
```

**Comment by manics** [comment]
BTW, I've just realised https://github.com/openmicroscopy/omero-server-docker/blob/master/requirements.yml#L21 means this can't be tagged as production until https://github.com/openmicroscopy/ansible-role-omero-server/pull/12 is tagged

File: 60-database.sh (hunk 4 above)
Comment (manics) [inline 60-database.sh]: No, `CONFIG_omero_db_host` will already be converted into omero.db.host by `50-config.py`- I'll update the comment at the top of the file. The only reason for special handling here is that the default of `omero.db.host=localhost` will never work, instead it defaults to `db` which should be set by `docker run --link ...`.

File: 50-config.py (hunk 5 above)
Comment (manics) [inline 50-config.py]: I know. Do you have another suggestion?

File: README.md (hunk 6 above)
Comment (manics) [inline README.md]: I didn't want to repeat all the `-e CONFIG_omero_db_*` params from the previous example, I'll try and make that clearer

File: 50-config.py (hunk 5 above)
Comment (joshmoore) [inline 50-config.py]: Not sure how evil this is but:

```
$ env "a.b=1" python -c "import os; print os.environ['a.b']"
1
$ docker run -ti --rm -e a.b=1 centos python -c "import os; print os.environ['a.b']"
1
```

**Comment by manics** [comment]
oops...not sure what happened...

**Comment by manics** [comment]
I had a typo, accidently did git push ... -d.

**Comment by joshmoore** [comment]
> I had a typo, accidently did git push ... -d.

That's kinda scary. Good to know.

**Comment by manics** [comment]
Updated in response to @joshmoore's comments.

In addition I've switched from loading `.omero` files from `/config/` to `/opt/omero/server/config/` as used in https://github.com/openmicroscopy/ansible-role-omero-server/tree/2.0.0-m1

One issue you may want to consider before ansible-role-omero-server is officially tagged `2.0.0`: the configuration script `/opt/omero/server/config/omero-server-config-update.sh` is in the same directory that's used for `.omero` files `/opt/omero/server/config` so you can mount `.omero` files but not the entire directory.

File: 50-config.py (hunk 5 above)
Comment (manics) [inline 50-config.py]: Unfortuntately it's also used in the bash script`60-database.sh`

**Comment by joshmoore** [comment]
Discussion now points to:

 * merging in order:
  - https://github.com/ome/omego/pull/104
  - https://github.com/openmicroscopy/ansible-role-omero-server/pull/12
  - this PR
 * and then considering omero-server the current candidate production image.

**Comment by manics** [comment]
--depends on https://github.com/ome/omego/pull/108
- This needed to allow installations without a DB to proceed. You can temporarily test this using https://github.com/manics/omero-server-docker/commit/34b1b073930b99ff200af690b5b2af1a15ffb70c

**Comment by joshmoore** [comment]
No objections from my side to having uid and gid be ARGs, but in general, then, you could also subclass the image and `usermod`. More critical from my POV is: what's the expectation when pulling the image from hub?

To your question in particular: I haven't run into a situation where I _can't_ chown the directory with a particular UID before mounting but I could imagine not _wanting_ to if that were to put something in particular at risk. (i.e. I don't trust the 1000 user on the host system in question)

**Comment by manics** [comment]
@ dpwrussell I agree with you. The correct way to solve this is to treat the omero-server Docker filesystem as readonly with the exception of defined directories for logs and temp-files which will need to be mounted as volumes. Then you can simply use `docker run --user UID ...` to set the user OMERO runs as.

The purpose of the last commit is to at make the default `omero-server` UID well-known as opposed to using whatever `useradd` picks as the UID.

**Comment by manics** [comment]
Though what would be even better is if all OMERO logs went to stdout/stderr, so just `/tmp` would need to be writeable.

**Comment by joshmoore** [comment]
`latest` is now 5.4. Shall we go with the flow and this repo matches the milestones of the 5.4 series with the intent of releasing this as production with 5.4.0?

cc: @dpwrussell @manics @sbesson @jburel

File: README.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+Additional configuration files for OMERO can be provided by mounting files into `/opt/omero/server/config/`.
+Files will be loaded with `omero load`.
+For example:
+
+    docker run -d -v
+        /config/extra.omero:/opt/omero/server/config/extra.omero:ro
+        openmicroscopy/omero-server
+
+Parameters required for initialising the server such as database configuration *must* be set using environment variables.
```
Comment (mtbc) [inline README.md]: initializing

Comment (manics) [inline README.md]: Changed

File: Dockerfile (hunk 3 above)
Comment (joshmoore) [inline Dockerfile]: I'm assuming this has everyone's implicit :thumbsup: but now with 5.4 replacing 5.3

**Comment by joshmoore** [comment]
Re-ran travis to get this green. Merging, untagged. I'd suggest when `5.4.0-m1` is tagged, that someone create a commit changing `latest` to `v5.4.0-m1` and tag that commit which should then be built on hub.

cc: @mtbc

File: 50-config.py (hunk 5 above)
Comment (joshmoore) [inline 50-config.py]: Continuing discussion under #3

**Pull request merged**